File: providers/Elasticsearch/src/MetadataCache.php
<?php

namespace HawaiianSearch;

class MetadataCache {
    private ElasticsearchClient $esClient;
    private int $maxSize;
    private int $flushInterval; // in seconds
    private array $cache = [];
    private float $lastFlushTime;
    private bool $dryrun;

    public function __construct(ElasticsearchClient $esClient, array $options = []) {
        $this->esClient = $esClient;
        $this->maxSize = isset($options['maxSize']) ? $options['maxSize'] : 1000;
        $this->flushInterval = isset($options['flushInterval']) ? $options['flushInterval'] : 60;
        $this->dryrun = isset($options['dryrun']) ? $options['dryrun'] : false;
        $this->lastFlushTime = microtime(true);
    }

    public function set(string $hash, array $metadata): void {
        $this->cache[$hash] = $metadata;
        if (!$this->dryrun && (count($this->cache) >= $this->maxSize || (microtime(true) - $this->lastFlushTime) >= $this->flushInterval)) {
            $this->flush();
        }
    }

    public function get(string $hash): ?array {
        // For simplicity in this context, we don't read from the index here.
        // The cache is write-through. The main CorpusScanner will pre-warm it if needed.
        return $this->cache[$hash] ?? null;
    }

    public function flush(): void {
        if (empty($this->cache)) {
            return;
        }

        echo "Flushing metadata cache (" . count($this->cache) . " items)..." . PHP_EOL;
        try {
            $this->esClient->updateMetadata($this->cache);
            $this->cache = []; // Clear the cache after successful flush
            $this->lastFlushTime = microtime(true);
        } catch (\Exception $e) {
            echo "⚠️ Failed to flush metadata cache: " . $e->getMessage() . PHP_EOL;
        }
    }

    // Call this at the end of the script to ensure any remaining items are saved.
    public function finish(): void {
        $this->flush();
    }
}

File: providers/Elasticsearch/python/metadata_cache.py
from collections import OrderedDict
import time

class MetadataCache:
    def __init__(self, es_client, max_size=10000, flush_interval=60):
        self.es = es_client
        self.cache = OrderedDict()
        self.dirty = set()  # hashes with modified values
        self.max_size = max_size
        self.last_flush = time.time()
        self.flush_interval = flush_interval  # in seconds

    def get(self, h):
        if h in self.cache:
            self.cache.move_to_end(h)
            return self.cache[h]

        meta = self.es.get_metadata(h) or {}
        self.cache[h] = meta
        return meta

    def set(self, h, value):
        self.cache[h] = value
        self.dirty.add(h)
        self.cache.move_to_end(h)

        if len(self.cache) > self.max_size:
            self.evict()

        now = time.time()
        if now - self.last_flush >= self.flush_interval:
            self.flush()

    def evict(self):
        old_h, old_val = self.cache.popitem(last=False)
        if old_h in self.dirty:
            self.es.update_metadata({old_h: old_val})
            self.dirty.remove(old_h)

    def flush(self):
        updates = {h: self.cache[h] for h in self.dirty}
        if updates:
            self.es.update_metadata(updates)
            self.dirty.clear()
            self.last_flush = time.time()

    def finish(self):
        self.flush()
